### ISSUE TITLE
[QST] API design

### ISSUE THREAD

**Issue by hadi-gharibi** [NVIDIA/NVTabular#173]
I have some questions regarding the API and how it designed.

- Why does the dataset have different classes instead of having one main class with some **_classmethod_** like **_read_x_** and **_write_x_**?

- Why ops have this complicated structure(I couldn't understand them honestly), what was wrong with Transformer structure of Sklearn? Is it possible to have something like Sklearn Pipeline here as well?

**Comment by rjzamora** [comment]
Thanks for the questions @hadi-gharibi !  Are you referring to the the latest version of nvtabular in this repo, or the 0.1.1 (alpha) release?  Note that the current development version includes a few notable changes to the API.  So, if your first question is based on previously released/documented code, the answer is probably “this has been changed”.

>Why does the dataset have different classes instead of having one main class with some classmethod like read_x and write_x?

The user is only meant to interact with one main `Dataset` class, and we want this interaction to be as simple as possible: `Dataset(data_source, <optional-args>)`.  It doesn’t really make much sense to expose a `Dataset.read_x|write_x` method, because the only purpose of this class is to translate an arbitrary data source into an object that a separate `Workflow` object can process.  In other words, there **shouldn’t** be a scenario where the user would need to call something like `read_parquet` *explicitly*, because the `Workflow` will do this if/when needed.

Perhaps you are asking why we use an underlying `DatasetEngine` (rather than just adding a new method for each format type)?   If so, this was just a design preference,  but it’s also a detail that *should* be invisible to the user.

A bit more info about `Dataset`… The ultimate goal of this class is to translate various on-disk and in-memory formats into an internal dask_cudf DataFrame, so that the workflow can process the data in a *scalable* way.  We ultimately want `Dataset` to be a very thin wrapper around the various dask_cudf IO methods (read_parquet, read_csv, read_orc, from_pandas, form_dask_dataframe, from_cudf, etc…).  Since the user is not meant to do anything with a `Dataset` object besides pass it to a `Workflow`,  it probably makes sense to avoid adding a bunch of class methods since it wold just add an extra step for the user.


>Why ops have this complicated structure(I couldn't understand them honestly), what was wrong with Transformer structure of Sklearn? Is it possible to have something like Sklearn Pipeline here as well?


We are very interested to get feedback on the “ops” user experience, so I’m glad to see someone start a discussion around this topic.  Could you clarify if you find it confusing to add an op to a workflow, or if you find it confusing to implement an op (or both)?

My understanding of the sklearn transformer structure is that the transformation requires two steps (1) `fit`, and (2) `transform`.  We could certainly adopt these terms by changing `update_stats` to `fit`, and `apply` to `fit_transforms`.  We could also use the fit/transform vocabulary within the actual op implementations.   Overall, we are doing the same thing, but happen to be using a stats/apply vocabulary instead of fit/transform.

If you are asking why nvtabular doesn’t make it possible to do something like `op.fit_transform(dataset)`,  then the answer is that the primary goal of nvtabular is to handle datasets that are much larger than GPU memory.  That is, we assume that it is critical to minimize the number of global passes through the dataset (including **all** transformations).  In order to do this, we use Dask to construct a graph of all required “fit” and “transform” tasks, and try to fuse these tasks together whenever possible.



Note that we are very much open to feedback and suggestions.  Overall, we aim to make the user experience as pleasant as possible :)

**Comment by hadi-gharibi** [comment]
That was a really nice and detailed explanation of the situation. For the first question, I was talking about why not having one new method for each format, because I thought it's nice to see something familiar(Although logically it's not logically completely right). But as you said it's just a design choice.
About the second question, as you said, Sklean custom transformation need to base classes, and use have to developed 
 `fit` and `transform` and it will take care of the rest of the stuff(e.g setters and getters). `fit_transform` is something that you would get as a bonus because of the base class. Then you could use the `Pipeline` class to pipe all of those steps together and from there you could use `fit` and `transform` functions as well(Which I think here `Workflow` is doing that).
`fit ` function could still be lazy, for example doing nothing for those `ZeroFill`, but for something like `Normalize` we could save mean and std as `self.mean` and `self.std` in our `fit` phase, and actually compute them at `transform`. I think this format would not only more familiar to a bigger community but also help us to develop more ops by only have some knowledge of `cudf_dask`.
BTW, do you have some contribution guide, that we could use?
